test/framework/machinehealthcheck_helpers.go

```diff
@@ -0,0 +1,169 @@
+	Expect(input.MachineHealthCheckName).ToNot(BeEmpty(), "Invalid argument. input.MachineHealthCheckName can't be empty when calling WaitForMachineHealthCheckToDetectUnhealthyNodeCondition")
+
+	fmt.Fprintf(GinkgoWriter, "Checking if MachineHealthCheck machines are initially healthy\n")
+	mhc := &clusterv1.MachineHealthCheck{}
+	Eventually(func() bool {
+		Expect(input.ClusterProxy.GetClient().Get(ctx, client.ObjectKey{Namespace: input.Cluster.Namespace, Name: input.MachineHealthCheckName}, mhc)).To(Succeed())
+		return len(mhc.Spec.UnhealthyConditions) > 0 && mhc.Status.ExpectedMachines > 0 && mhc.Status.CurrentHealthy == mhc.Status.ExpectedMachines
+	}, intervals...).Should(BeTrue())
+
+	fmt.Fprintf(GinkgoWriter, "Patching MachineHealthCheck unhealthy condition to worker nodes\n")
+	unhealthyNodeCondition := corev1.NodeCondition{
+		Type:               mhc.Spec.UnhealthyConditions[0].Type,
+		Status:             mhc.Spec.UnhealthyConditions[0].Status,
+		LastTransitionTime: metav1.Time{Time: time.Now()},
+	}
+	for _, md := range input.MachineDeployments {
+		selectorMap, err := metav1.LabelSelectorAsMap(&md.Spec.Selector)
+		Expect(err).ToNot(HaveOccurred())
+
+		ms := &clusterv1.MachineSetList{}
+		Expect(input.ClusterProxy.GetClient().List(ctx, ms, client.InNamespace(input.Cluster.Namespace), client.MatchingLabels(selectorMap))).To(Succeed())
+		Expect(len(ms.Items)).NotTo(Equal(0))
+
+		machineSet := ms.Items[0]
+		selectorMap, err = metav1.LabelSelectorAsMap(&machineSet.Spec.Selector)
+		Expect(err).ToNot(HaveOccurred())
+		machines := &clusterv1.MachineList{}
+		Expect(input.ClusterProxy.GetClient().List(ctx, machines, client.InNamespace(machineSet.Namespace), client.MatchingLabels(selectorMap))).To(Succeed())
+
+		for _, machine := range machines.Items {
+			node := &corev1.Node{}
+			Expect(input.ClusterProxy.GetWorkloadCluster(ctx, input.Cluster.Namespace, input.Cluster.Name).GetClient().Get(ctx, types.NamespacedName{Name: machine.Status.NodeRef.Name, Namespace: machine.Status.NodeRef.Namespace}, node)).To(Succeed())
+			patchHelper, err := patch.NewHelper(node, input.ClusterProxy.GetWorkloadCluster(ctx, input.Cluster.Namespace, input.Cluster.Name).GetClient())
+			Expect(err).ToNot(HaveOccurred())
+			node.Status.Conditions = append(node.Status.Conditions, unhealthyNodeCondition)
+			Expect(patchHelper.Patch(ctx, node)).To(Succeed())
+		}
+	}
+
+	fmt.Fprintf(GinkgoWriter, "Patching MachineHealthCheck to initiate reconcile\n")
+	Expect(input.ClusterProxy.GetClient().Get(ctx, client.ObjectKey{Namespace: input.Cluster.Namespace, Name: input.MachineHealthCheckName}, mhc)).To(Succeed())
+	patchHelper, err := patch.NewHelper(mhc, input.ClusterProxy.GetClient())
+	Expect(err).ToNot(HaveOccurred())
+	mhc.Labels["e2e.x-k8s.io"] = "triggered"
+	Expect(patchHelper.Patch(ctx, mhc)).To(Succeed())
```

Comment:
Why do we need to patch MHC and force a reconciliation? Don't we subscribe to the remote cluster's node events?

Reply:
I don't think it picks up node changes right away, has to wait until next reconcile. So I had to trigger it like this.

Reply:
That seems odd, based on the work that has been done on informers in the MHC original codebase.

cc @JoelSpeed

Reply:
Assuming the MHC has been reconciled already, there should be a remote cluster watcher so assuming a Node object is updated on the "remote cluster", then MHC should reconcile automatically. Manually triggering it shouldn't be necessary.

I haven't yet looked into what this PR is doing but I'll try and find some time shortly

Reply:
Ok i've found the issue, the map function ends up checking labels using `hasMatchingLabels`, comparing the MHC selector and machine labels. I thought we had agreed empty selectors were allowed but this map function currently does the opposite. @vincepri What did you think we had agreed on the empty selector front, should we consider this a bug and fix it up?

https://github.com/kubernetes-sigs/cluster-api/blob/23a9ec341de33c399c542558c22046cb413082d5/controllers/machine_helpers.go#L60-L63

In the mean time, @sedefsavas you should add a selector to the MHCs that you are creating which will resolve the issue and in theory should make the tests more robust in the future

Reply:
To confirm, when doing the selection in the opposite direction, MHC -> Machine, we allow an empty selector and match all nodes. So this an MHC bug IMO.

https://github.com/kubernetes-sigs/cluster-api/blob/23a9ec341de33c399c542558c22046cb413082d5/controllers/machinehealthcheck_targets.go#L180-L197

Reply:
Thanks @JoelSpeed. I modified it to use selectors but will have another PR for that.
For this inconsistency, IMO we should create an issue.

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/3006